### ISSUE TITLE
Failing to Patch Method due to System.FormatException IL Code

### ISSUE THREAD

**Issue by vliberio** [pardeike/Harmony#185]
I apologize for bothering you. I am using Harmony 1.2.0.1 on Cities: Skylines on .NET 3.5 on Windows 10 64-bit. I'm trying to use your library to prefix and postfix a method (NetTool.CreateNode in C:S:) and I can't seem to figure out what I am doing wrong. I have crawled through the other issues and Discord.

**Original Method (according to ILSpy):**
Signature:
```cs
public static ToolErrors CreateNode(NetInfo info, ControlPoint startPoint, ControlPoint middlePoint, ControlPoint endPoint, FastList<NodePosition> nodeBuffer, int maxSegments, bool test, bool testEnds, bool visualize, bool autoFix, bool needMoney, bool invert, bool switchDir, ushort relocateBuildingID, out ushort firstNode, out ushort lastNode, out ushort segment, out int cost, out int productionRate);
```
Code:
[CreateNodeCS.txt](https://github.com/pardeike/Harmony/files/3168557/CreateNodeCS.txt)
[CreateNodeIL.txt](https://github.com/pardeike/Harmony/files/3168558/CreateNodeIL.txt)

**Patch Code:***
```cs
    [HarmonyPatch(typeof(NetTool))]
    [HarmonyPatch("CreateNode")]
    [HarmonyPatch(new Type[] { typeof(NetInfo), typeof(ControlPoint), typeof(ControlPoint), typeof(ControlPoint), typeof(FastList<NodePosition>), typeof(int), typeof(bool), typeof(bool), typeof(bool), typeof(bool), typeof(bool), typeof(bool), typeof(bool), typeof(ushort), typeof(ushort), typeof(ushort), typeof(ushort), typeof(int), typeof(int)}, new ArgumentType[] { ArgumentType.Normal, ArgumentType.Normal, ArgumentType.Normal, ArgumentType.Normal, ArgumentType.Normal, ArgumentType.Normal, ArgumentType.Normal, ArgumentType.Normal, ArgumentType.Normal, ArgumentType.Normal, ArgumentType.Normal, ArgumentType.Normal, ArgumentType.Normal, ArgumentType.Normal, ArgumentType.Ref, ArgumentType.Ref, ArgumentType.Ref, ArgumentType.Ref, ArgumentType.Ref })]
    public class ToolCreateNode
    {
        public static void Prefix(NetInfo info, ControlPoint startPoint, ControlPoint middlePoint, ControlPoint endPoint, FastList<NodePosition> nodeBuffer, int maxSegments, bool test, bool testEnds, bool visualize, bool autoFix, bool needMoney, bool invert, bool switchDir, ushort relocateBuildingID, ushort firstNode, ushort lastNode, ushort segment, int cost, int productionRate)
        {
            Debug.Log("Prefixing NetTool::CreateNode!");
        }

        public static void Postfix(NetInfo info, ControlPoint startPoint, ControlPoint middlePoint, ControlPoint endPoint, FastList<NodePosition> nodeBuffer, int maxSegments, bool test, bool testEnds, bool visualize, bool autoFix, bool needMoney, bool invert, bool switchDir, ushort relocateBuildingID, ushort firstNode, ushort lastNode, ushort segment, int cost, int productionRate)
        {
            Debug.Log("Postfixing NetTool::CreateNode!");
        }
    }
```

**Debug Info:**
[harmony.log.txt](https://github.com/pardeike/Harmony/files/3168564/harmony.log.txt)
```
[Error] Patching failed System.FormatException: Method NetTool.CreateNode(NetInfo, NetTool+ControlPoint, NetTool+ControlPoint, NetTool+ControlPoint, FastList`1[[NetTool+NodePosition, Assembly-CSharp, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null]], System.Int32, System.Boolean, System.Boolean, System.Boolean, System.Boolean, System.Boolean, System.Boolean, System.Boolean, System.UInt16, System.UInt16&, System.UInt16&, System.UInt16&, System.Int32&, System.Int32&) cannot be patched. Reason: Invalid IL code in (wrapper dynamic-method) NetTool:CreateNode_Patch2 (NetInfo,NetTool/ControlPoint,NetTool/ControlPoint,NetTool/ControlPoint,FastList`1<NetTool/NodePosition>,int,bool,bool,bool,bool,bool,bool,bool,uint16,uint16&,uint16&,uint16&,int&,int&): IL_0005: stloc.s   178


  at Harmony.PatchFunctions.UpdateWrapper (System.Reflection.MethodBase original, Harmony.PatchInfo patchInfo, System.String instanceID) [0x00000] in <filename unknown>:0 
  at Harmony.PatchProcessor.Patch () [0x00000] in <filename unknown>:0 
  at Harmony.HarmonyInstance.<PatchAll>b__9_0 (System.Type type) [0x00000] in <filename unknown>:0 
  at Harmony.CollectionExtensions.Do[Type] (IEnumerable`1 sequence, System.Action`1 action) [0x00000] in <filename unknown>:0 
  at Harmony.HarmonyInstance.PatchAll (System.Reflection.Assembly assembly) [0x00000] in <filename unknown>:0 
  at CSM.CSM..ctor () [0x00000] in <filename unknown>:0
```

With my 3 and 3/4 minutes worth of experience in IL code, I understand that it's trying to put 0 into another ToolErrors local variable and dying. I can't seem to figure out why it's doing that or even where that ToolErrors local variable 178 or the code messing with it came from. I would appreciate any help you can give me.

**Comment by pardeike** [comment]
Is `ToolErrors` a struct? Harmony 1.2.* can’t handle methods returning structs of certain sizes because it does not support certain optimisations done in .NET method callings for such methods.

**Comment by pardeike** [comment]
Ah, found the clue in your IL log:
`valuetype ToolBase/ToolErrors`

See discussion of “this pointer” here: https://github.com/dotnet/coreclr/blob/master/Documentation/botr/clr-abi.md

**Comment by pardeike** [comment]
Fixed in the upcoming 2.0 beta
